### ISSUE TITLE
@JsonAdapter + JsonDeserializer make JsonSerializer fail

### ISSUE THREAD

**Issue by DoneSpeak** [google/gson#1783]
I want to deserialize the string to list so I use `@JsonAdapter + JsonDeserializer` to solve the problem, but when I try to serialize the List, I get a `{}`.

```xml
<dependency>
    <groupId>com.google.code.gson</groupId>
    <artifactId>gson</artifactId>
    <version>2.8.5</version>
</dependency>
```

```java
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.stream.Collectors;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import com.google.gson.JsonPrimitive;
import com.google.gson.annotations.JsonAdapter;
import lombok.Data;

public class Temp {

    @Data
    public static class DataValue {
        @JsonAdapter(ListDeserializer.class)
        private List<String> list;
    }

    public static class ListDeserializer implements JsonDeserializer<List<String>> {

        @Override
        public List<String> deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
            throws JsonParseException {
            if(json.isJsonNull()) {
                return null;
            }
            if (json.isJsonNull()) {
                return null;
            }
            if (json.isJsonArray()) {
                Iterator<JsonElement> iter = json.getAsJsonArray().iterator();
                final List<String> result = new ArrayList<>();
                iter.forEachRemaining(e -> result.add(e.getAsString()));
                return result;
            }
            if (json.isJsonPrimitive()) {
                JsonPrimitive jsonPrimitive = (JsonPrimitive) json;
                String str = jsonPrimitive.getAsString();
                return Arrays.asList(str.split(","));
            }
            return null;
        }
    }

    public static void main(String[] args) {
        GsonBuilder gsonBuilder = new GsonBuilder();
        // gsonBuilder.setPrettyPrinting();
        Gson gson = gsonBuilder.create();

        String json = "{\"list\":\"a,b,c,d,e,f\"}";
        DataValue dataValue = gson.fromJson(json, DataValue.class);
        // Temp.DataValue(list=[a, b, c, d, e, f])
        System.out.println(dataValue);
        // {"list":{}}
        System.out.println(gson.toJson(dataValue));
    }
}
```

**Comment by Marcono1234** [comment]
It looks like this describes the same problem as #1326 (though that issue is not as detailed).

The underlying issue is similar to #1028: Because `ListDeserializer` does not implement the `JsonSerializer` interface the `TreeTypeAdapter` (internal class) tries to get the delegate using `Gson.getDelegateAdapter(TypeAdapterFactory, TypeToken)`. This method erroneously thinks that the adapter factory was created using a `@JsonAdapter` placed on the class (instead of the field which is the case in your code) and therefore skips nearly all built-in adapters, falling back to the reflection based serialization. However not even the backing array field of the list is serialized because the `TreeTypeAdapter` looked up the adapter for the interface `List` (instead of the runtime class of the value, i.e. `java.util.Arrays.ArrayList`) which does not have any fields.

**Comment by Marcono1234** [comment]
It appears this has been fixed by #2435; the output is now:
```json
{"list":["a","b","c","d","e","f"]}
```
(though I omitted the Lombok `@Data` annotation for testing)

I am therefore going to close this issue. Though please let us know if you can still reproduce this with the latest Gson code from the `main` branch, respectively the next upcoming Gson version (2.11.0 probably).
